File: javav2/example_code/emr/src/main/java/aws/example/emr/AddSteps.java
//snippet-sourcedescription:[AddSteps.java demonstrates how to add new steps to a running cluster.]
//snippet-keyword:[AWS SDK for Java v2]
//snippet-keyword:[Code Sample]
//snippet-keyword:[Amazon EMR]
//snippet-sourcetype:[full-example]

/*
   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   SPDX-License-Identifier: Apache-2.0
*/

package aws.example.emr;

// snippet-start:[emr.java2._add_steps.import]
import software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.emr.EmrClient;
import software.amazon.awssdk.services.emr.model.AddJobFlowStepsRequest;
import software.amazon.awssdk.services.emr.model.EmrException;
import software.amazon.awssdk.services.emr.model.HadoopJarStepConfig;
import software.amazon.awssdk.services.emr.model.StepConfig;
// snippet-end:[emr.java2._add_steps.import]

/**
 * Before running this Java V2 code example, set up your development environment, including your credentials.
 *
 * For more information, see the following documentation topic:
 *
 * https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/get-started.html
 */

public class AddSteps {

    public static void main(String[] args) {

        final String usage = "\n" +
            "Usage: " +
            "   <jar> <myClass> <jobFlowId> \n\n" +
            "Where:\n" +
            "   jar - A path to a JAR file run during the step. \n\n" +
            "   myClass - The name of the main class in the specified Java file. \n\n" +
            "   jobFlowId - The id of the job flow. \n\n" ;

        if (args.length != 3) {
            System.out.println(usage);
            System.exit(1);
        }

        String jar = args[0] ;
        String myClass = args[1] ;
        String jobFlowId = args[2] ;
        Region region = Region.US_WEST_2;
        EmrClient emrClient = EmrClient.builder()
            .region(region)
            .credentialsProvider(ProfileCredentialsProvider.create())
            .build();

        addNewStep(emrClient, jobFlowId, jar, myClass);
        emrClient.close();
    }

    // snippet-start:[emr.java2._add_steps.main]
    public static void addNewStep(EmrClient emrClient, String jobFlowId, String jar, String myClass) {

        try {
            HadoopJarStepConfig jarStepConfig = HadoopJarStepConfig.builder()
                .jar(jar)
                .mainClass(myClass)
                .build();

            StepConfig stepConfig = StepConfig.builder()
                .hadoopJarStep(jarStepConfig)
                .name("Run a bash script")
                .build();

            AddJobFlowStepsRequest jobFlowStepsRequest = AddJobFlowStepsRequest.builder()
                .jobFlowId(jobFlowId)
                .steps(stepConfig)
                .build();

            emrClient.addJobFlowSteps(jobFlowStepsRequest);
            System.out.println("You have successfully added a step!");

        } catch (EmrException e) {
            System.err.println(e.getMessage());
            System.exit(1);
        }
    }
    // snippet-end:[emr.java2._add_steps.main]
}

File: kotlin/services/emr/src/main/kotlin/com/kotlin/emr/AddSteps.kt
// snippet-sourcedescription:[AddSteps.kt demonstrates how to add new steps to a running cluster.]
// snippet-keyword:[AWS SDK for Kotlin]
// snippet-keyword:[Amazon EMR]

/*
   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
   SPDX-License-Identifier: Apache-2.0
*/

package com.kotlin.emr

// snippet-start:[erm.kotlin.add_steps.import]
import aws.sdk.kotlin.services.emr.EmrClient
import aws.sdk.kotlin.services.emr.model.AddJobFlowStepsRequest
import aws.sdk.kotlin.services.emr.model.HadoopJarStepConfig
import aws.sdk.kotlin.services.emr.model.StepConfig
import kotlin.system.exitProcess
// snippet-end:[erm.kotlin.add_steps.import]

/**
Before running this Kotlin code example, set up your development environment,
including your credentials.

For more information, see the following documentation topic:
https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/setup.html
*/

suspend fun main(args: Array<String>) {

    val usage = """
        Usage:
            <jar> <myClass> <jobFlowId>

        Where:
            jar - A path to a JAR file run during the step. 
            myClass - The name of the main class in the specified Java file. 
            jobFlowId - The id of the job flow.
        """

    if (args.size != 3) {
        println(usage)
        exitProcess(0)
    }

    val jar = args[0]
    val myClass = args[1]
    val jobFlowId = args[2]
    addNewStep(jobFlowId, jar, myClass)
}

// snippet-start:[erm.kotlin.add_steps.main]
suspend fun addNewStep(jobFlowIdVal: String?, jarVal: String?, myClass: String?) {

    val jarStepConfig = HadoopJarStepConfig {
        jar = jarVal
        mainClass = myClass
    }

    val stepConfig = StepConfig {
        hadoopJarStep = jarStepConfig
        name = "Run a bash script"
    }

    val request = AddJobFlowStepsRequest {
        jobFlowId = jobFlowIdVal
        steps = listOf(stepConfig)
    }

    EmrClient { region = "us-west-2" }.use { emrClient ->
        emrClient.addJobFlowSteps(request)
        println("You have successfully added a step!")
    }
}
// snippet-end:[erm.kotlin.add_steps.main]
